tutorials/haskell/bdd-in-haskell-with-hspec/tutorial.md

```diff
@@ -0,0 +1,523 @@
+rather than expectations. This means that if you have some piece of code,
+which must hold on a property that should always be true, you can use
+QuickCheck to feed that property, with as much arbitrary inputs
+as possible, to check it really holds.
+
+For educative purposes we'll add a `fromDigits` function. This function should
+do the opposite of what `toDigits` does. It should take a list of digits and
+convert it to a number.
+
+Let's start with some code as if we don't have QuickCheck around.
+We can start by adding this code to the `describe "toDigits"` section
+in the `LuhnSpec` file:
+
+```haskell
+    it "holds on: x == fromDigits(toDigits x)" $ do
+      12345 `shouldBe` fromDigits (toDigits 123456)
+```
+
+And then add this simple implementation to the `Luhn.Internal` module.
+Remember to expose this function in the `module` clause:
+
+```haskell
+fromDigits :: [Integer] -> Integer
+fromDigits _ = 12345
+```
+
+Spec passes and everyone is happy. Now let's replace the spec we just
+wrote with a QuickCheck a property instead:
+
+```haskell
+    it "holds on the property x == fromDigits(toDigits x)" $ do
+      property $ \x -> x == (fromDigits . toDigits) x
+```
+
+Here we are creating a property. The property receives a function that will receive
+arbitrary values of the inferred type, and must return a `Bool`. In this case, `x` is
+going to be of type `Integer` and we use it to assert the property.
+
+Under the hood QuickCheck will feed the property function with many
+arbitrary values, and the property checks if and only if, for all cases,
+the function returns `True`.
+
+After running specs and checking the property doesn't hold, we add a suitable
+implementation:
+
+```haskell
+fromDigits :: [Integer] -> Integer
+fromDigits = sum . zipWith (*) [10 ^ i | i <- [0..]] . reverse
+```
+
+It didn't pass. What's going on? Let's check the Hspec display:
+
+```
+test/LuhnSpec.hs:22:
+  1) Luhn.toDigits holds on: x == fromDigits(toDigits x)
+       Falsifiable (after 3 tests):
+       -1
+```
+
+Aha! The property doesn't hold for negative numbers, and that's all right,
+after all, this was the expected behaviour. The `toDigits` function returns an
+empty list for numbers lower or equal to zero. So we cannot recover the same
+number back with the `fromDigits` function.
+
+The solution is to prevent QuickCheck from feeding negative numbers.
+Haskell has a type called `Natural` which is an instance of `Integral`, but to
+access this type we need to import `Numeric.Natural` first in our `LuhnSpec` file:
+
+```haskell
+import Numeric.Natural
+```
+
+We'll rewrite our spec once more:
+
+```haskell
+      property $ \x -> let x' = toInteger (x :: Natural) in
+        x' == (fromDigits . toDigits) x'
+```
+
+We are forcing the type of `x` to be a `Natural` and later convert `Natural` to
+`Integer` to be able to feed our functions. This time around, it works like a
+charm.
+
```

Comment:
You can avoid using `Natural` and use QuickCheck's built-in helper wrapper [`NonNegative`](https://hackage.haskell.org/package/QuickCheck-2.8.2/docs/Test-QuickCheck.html#t:NonNegative). This achieves two goals in the context of the tutorial:

1. You don't bring new data type `Natural` in scope as it has little to do with your orginal intention to use `Integral` and is used rather as a means to make QuickCheck generate correct inputs. Also note that since you have figured out now that negative numbers should not inhabit type of argument of `toDigits` (and `fromDigits` too), maybe it's better to use `Natural` (or `Word`) to capture this on type level from the beginning.

2. By bringing `NonNegative` helper into the scope, you can tell something about similar helpers like `NonEmptyList`, etc. These are generally useful and worth mentioning.

As a side note, you could also mention here that `(==>)` allows to specify arbitrary prerequisites for your properties, such as `x > 0 ===> x == (fromDigits . toDigits) x` without type-acrobatics at all.